x/auction/keeper/auctions.go

```diff
@@ -0,0 +1,337 @@
+// TODO passing bid and lot is weird when only one needed
+func (k Keeper) PlaceBid(ctx sdk.Context, auctionID uint64, bidder sdk.AccAddress, bid sdk.Coin, lot sdk.Coin) sdk.Error {
+
+	// get auction from store
+	auction, found := k.GetAuction(ctx, auctionID)
+	if !found {
+		return sdk.ErrInternal("auction doesn't exist")
+	}
+
+	// validate
+	if ctx.BlockTime().After(auction.GetEndTime()) {
+		return sdk.ErrInternal("auction has closed")
+	}
+	if auction.GetBid().Denom != bid.Denom {
+		return sdk.ErrInternal("bid has incorrect denom")
+	}
+	if auction.GetLot().Denom != lot.Denom {
+		return sdk.ErrInternal("lot has incorrect denom")
+	}
+
+	// place bid
+	var err sdk.Error
+	var a types.Auction
+	switch auc := auction.(type) {
+	case types.ForwardAuction:
+		a, err = k.PlaceBidForward(ctx, auc, bidder, bid)
+		if err != nil {
+			return err
+		}
+	case types.ReverseAuction:
+		a, err = k.PlaceBidReverse(ctx, auc, bidder, lot)
+		if err != nil {
+			return err
+		}
+	case types.ForwardReverseAuction:
+		a, err = k.PlaceBidForwardReverse(ctx, auc, bidder, bid, lot)
+		if err != nil {
+			return err
+		}
+	default:
+		panic(fmt.Sprintf("unrecognized auction type: %T", auction))
+	}
+
+	// store updated auction
+	k.SetAuction(ctx, a)
+	return nil
+}
+
+func (k Keeper) PlaceBidForward(ctx sdk.Context, a types.ForwardAuction, bidder sdk.AccAddress, bid sdk.Coin) (types.ForwardAuction, sdk.Error) {
+	// Valid New Bid
+	if bid.Denom != a.Bid.Denom {
+		return a, sdk.ErrInternal("bid denom doesn't match auction")
+	}
+	if !a.Bid.IsLT(bid) { // TODO add minimum bid size
+		return a, sdk.ErrInternal("bid not greater than last bid")
+	}
+
+	// Move Coins
+	increment := bid.Sub(a.Bid)
+	bidAmtToReturn := a.Bid
+	if bidder.Equals(a.Bidder) { // catch edge case of someone updating their bid with a low balance
+		bidAmtToReturn = sdk.NewInt64Coin(a.Bid.Denom, 0)
+	}
+	err := k.supplyKeeper.SendCoinsFromAccountToModule(ctx, bidder, types.ModuleName, sdk.NewCoins(bidAmtToReturn.Add(increment)))
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToAccount(ctx, types.ModuleName, a.Bidder, sdk.NewCoins(bidAmtToReturn))
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, a.Initiator, sdk.NewCoins(increment)) // increase in bid size is burned
+	if err != nil {
+		return a, err
+	}
+	err = k.supplyKeeper.BurnCoins(ctx, a.Initiator, sdk.NewCoins(increment))
+	if err != nil {
+		return a, err
+	}
+
+	// Update Auction
+	a.Bidder = bidder
+	a.Bid = bid
+	// increment timeout
+	a.EndTime = earliestTime(ctx.BlockTime().Add(types.DefaultBidDuration), a.MaxEndTime)
+
+	return a, nil
+}
+func (k Keeper) PlaceBidForwardReverse(ctx sdk.Context, a types.ForwardReverseAuction, bidder sdk.AccAddress, bid sdk.Coin, lot sdk.Coin) (types.ForwardReverseAuction, sdk.Error) {
+	// Validate New Bid // TODO min bid increments, make validation code less confusing
+	if !a.Bid.IsEqual(a.MaxBid) {
```

Comment:
Replacing this check with `a.Phase == AuctionPhase.ForwardPhase` would make this more readable. Auction starts `ForwardPhase` and is updated to `ReversePhase` at the end of the function `if bid.IsEqual(a.MaxBid)`.

Reply:
I've added a auction method `IsReversePhase` to make things clearer.
Thought this might be better than an enum on the struct as the phase can be determined from the fields that are already present.